convert2rhel/subscription.py

```diff
@@ -279,26 +288,89 @@ def args(self):
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus looks to be installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8:
+            * RHEL9:
+                * dbus looks to be installed and run by default
+                * python3-dbus-1.2.18 is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html#registerserver
         """
-        if self.password:
-            loggerinst.debug(
-                "Calling command '%s %s'" % (self.cmd, " ".join(hide_secrets(self.args)))
-            )  # lgtm[py/clear-text-logging-sensitive-data]
-            output, ret_code = utils.run_cmd_in_pty(
-                [self.cmd] + self.args, expect_script=(("[Pp]assword: ", self.password + "\n"),), print_cmd=False
-            )
+        # Questions:
+        # * Should we check whether dbus is running?
+        # * If it's not running should we install and run it or should we fall back to
+        #   subscription-managr cli?
+        # * If we install it, do we need to revert that after the upgrade/rollback?
+        # * Should checking for DBUS be done in either system_info or checks.py?
+        # * What should the return value of this code be?
+
+        # DBus throws errors for pure empty dict and dbus-python doesn't like
+        # {"": ""} even though the /bin/dbus-send thinks that is okay.  Use
+        # this instead.
+        EMPTY_DBUS_DICT = dbus.Dictionary({}, signature="ss")
+
+        loggerinst.debug("Getting a handle to the system dbus")
+        system_bus = dbus.SystemBus()
+
+        if self.server_url:
+            connection_opts = {"host": self.server_url}
         else:
-            # Warning: Currently activation_key can only be specified on the CLI. This is insecure
-            # but there's nothing we can do about it for now. Once subscription-manager issue:
-            # https://issues.redhat.com/browse/ENT-4724 is implemented, we can change both password
-            # and activation_key to use a file-based approach to passing the secrets.
-            output, ret_code = utils.run_subprocess([self.cmd] + self.args, print_cmd=False)
+            connection_opts = EMPTY_DBUS_DICT
+
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start("C", dbus_interface="com.redhat.RHSM1.RegisterServer")
```

Comment:
I like the idea of moving the locale out to a separate file in case we do localize convert2rhel in the future. I'm a little hesitant about changing this to use a non-C locale for several reasons, though:

* convert2rhel isn't localized itself.  Emitting a few localized messages while everything else is unlocalized doesn't seem like a good experience.
* In the past I've had some people make the point that error messages, are better left unlocalized so that people can easily put the errors into google (or bug reports).  I'm not quite convinced either way about that but I think it's appropriate to punt on that question until we decide to localize convert2rhel itself.

Reply:
> I like the idea of moving the locale out to a separate file

Oh sorry, I did not mean a new file on its own, rather a simple variable declared at the top of the `__call__()` function, eg.
```python
    SUBMAN_LOCALE = "C"
```
and use it in all the calls; this way, whatever you want to do in the future, it's a one place change.

Reply:
I moved the locale constants out to their own file in this commit: f90e34f715f955fd4514f6bc07be070c6c16e786